server/apm/apm-server/apm-service/src/main/java/io/holoinsight/server/apm/server/service/impl/ApmMetricMaterializer.java

```diff
@@ -0,0 +1,148 @@
+    SCHEDULER.scheduleAtFixedRate(this::materialize, wait, INTERVAL, TimeUnit.MILLISECONDS);
+  }
+
+  private void materialize() {
+    long start = System.currentTimeMillis() / INTERVAL * INTERVAL - INTERVAL;
+    long end = start + INTERVAL;
+    Map<String, List<String>> indexGroups = new HashMap<>();
+    MetaDataDictValue cachedGroups = superCacheService.getSc().metaDataDictValueMap
+        .getOrDefault("global_config", new HashMap<>()).get("apm_materialized_groups");
+    if (cachedGroups != null) {
+      indexGroups = GsonUtils.fromJson(cachedGroups.getDictValue(),
+          new TypeToken<Map<String, List<String>>>() {}.getType());
+    }
+    QueryWrapper<TenantOps> wrapper = new QueryWrapper<>();
+    List<TenantOps> tenantOpsList = tenantOpsMapper.selectList(wrapper);
+    if (CollectionUtils.isNotEmpty(tenantOpsList)) {
+      Map<String, List<String>> finalIndexGroups = indexGroups;
+      tenantOpsList.forEach(tenantOps -> {
+        String tenant = tenantOps.getTenant();
+        Map<String, MetricDefine> metrics = metricsManager.getMetricDefines();
+        if (metrics != null) {
+          metrics.forEach((metricName, metricDefine) -> {
+            List<String> groups = finalIndexGroups.get(metricDefine.getIndex());
+            EXECUTOR.submit(() -> {
+              try {
+                log.info(
+                    "[apm] ready to materialize metric, tenant={}, metric={}, start={}, end={}, step={}, groups={}",
+                    tenant, metricName, start, end, STEP, groups);
+                MetricValues metricValues = apmMetricStorage.queryMetric(tenant, metricName,
+                    new Duration(start, end, STEP), null, groups);
+                log.info(
+                    "[apm] query metric success, tenant={}, metric={}, start={}, end={}, step={}, groups={}, series={}",
+                    tenant, metricName, start, end, STEP, groups,
+                    metricValues == null ? 0 : CollectionUtils.size(metricValues.getValues()));
+                WriteMetricsParam param = new WriteMetricsParam();
+                param.setTenant(tenant);
+                param.setPoints(new ArrayList<>());
+                AtomicInteger pointCnt = new AtomicInteger(0);
+                if (metricValues != null && CollectionUtils.isNotEmpty(metricValues.getValues())) {
+                  metricValues.getValues().forEach(value -> {
+                    value.getValues().forEach((t, v) -> {
+                      WriteMetricsParam.Point point = new WriteMetricsParam.Point();
+                      point.setMetricName(metricName);
+                      point.setTags(value.getTags());
+                      point.setTimeStamp(t);
+                      point.setValue(v);
+                      param.getPoints().add(point);
+                      pointCnt.incrementAndGet();
+                    });
+                  });
+                }
+                metricStorage.write(param).subscribe(unused -> {
```

Comment:
The 'metricStorage.write(param)' returns a `Mono` object which is something like a `ListenableFuture` of Guava.
The write operation itself is a async operation. It is better to use asynchronous programming instead of using a ThreadPool.
Check [Project Reactor](https://projectreactor.io/) for more details.

Reply:
Fixed.